executor.go

```diff
@@ -0,0 +1,253 @@
+package task
+
+import (
+	"context"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/sajari/fuzzy"
+
+	"github.com/go-task/task/v3/internal/compiler"
+	"github.com/go-task/task/v3/internal/logger"
+	"github.com/go-task/task/v3/internal/output"
+	"github.com/go-task/task/v3/internal/sort"
+	"github.com/go-task/task/v3/taskfile/ast"
+)
+
+type (
+	// An ExecutorOption is a functional option for an Executor
+	ExecutorOption func(*Executor)
+	// An Executor is a type that is used for processing and executing Taskfiles
+	Executor struct {
+		// Flags
+		Dir         string
+		Entrypoint  string
+		TempDir     TempDir
+		Force       bool
+		ForceAll    bool
+		Insecure    bool
+		Download    bool
+		Offline     bool
+		Timeout     time.Duration
+		Watch       bool
+		Verbose     bool
+		Silent      bool
+		AssumeYes   bool
+		AssumeTerm  bool // Used for testing
+		Dry         bool
+		Summary     bool
+		Parallel    bool
+		Color       bool
+		Concurrency int
+		Interval    time.Duration
+
+		// I/O
+		Stdin  io.Reader
+		Stdout io.Writer
+		Stderr io.Writer
+
+		// Internal
+		Taskfile           *ast.Taskfile
+		Logger             *logger.Logger
+		Compiler           *compiler.Compiler
+		Output             output.Output
+		OutputStyle        ast.Output
+		TaskSorter         sort.Sorter
+		UserWorkingDir     string
+		EnableVersionCheck bool
+
+		fuzzyModel *fuzzy.Model
+
+		concurrencySemaphore chan struct{}
+		taskCallCount        map[string]*int32
+		mkdirMutexMap        map[string]*sync.Mutex
+		executionHashes      map[string]context.Context
+		executionHashesMutex sync.Mutex
+	}
+	TempDir struct {
+		Remote      string
+		Fingerprint string
+	}
+)
+
+func NewExecutor(opts ...ExecutorOption) *Executor {
```

Comment:
Each of the functions in this file should have a documentation comment, do make them look pretty at https://pkg.go.dev/github.com/go-task/task/v3.